src/BuildQueue.cpp

```diff
@@ -255,6 +258,221 @@ namespace t2
+      for (const NodeInputFileData& input : prev_state->m_InputFiles)
+        JsonWriteValueString(msg, input.m_Filename);
+      JsonWriteEndArray(msg);
+
+      JsonWriteKeyName(msg, "dependency");
+      JsonWriteValueString(msg, "explicit");
+
+      JsonWriteEndObject(msg);
+
+      // Don't do any further checking for changes, it's going to be a lot of work figuring out which bits of old state
+      // correspond to which bits of new state, for very little benefit
+      return;
+    }
+
+    ReportChangedInputFiles(msg, prev_state->m_InputFiles, "explicit", digest_cache, stat_cache, sha_extension_hashes, sha_extension_hash_count);
+
+    if (node_data->m_Scanner)
+    {
+      HashTable<bool, kFlagPathStrings> implicitDependencies;
+      HashTableInit(&implicitDependencies, &thread_state->m_LocalHeap);
+
+      for (const FrozenFileAndHash& input : node_data->m_InputFiles)
+      {
+        // Roll back scratch allocator between scans
+        MemAllocLinearScope alloc_scope(&thread_state->m_ScratchAlloc);
+
+        ScanInput scan_input;
+        scan_input.m_ScannerConfig = node_data->m_Scanner;
+        scan_input.m_ScratchAlloc = &thread_state->m_ScratchAlloc;
+        scan_input.m_ScratchHeap = &thread_state->m_LocalHeap;
+        scan_input.m_FileName = input.m_Filename;
+        scan_input.m_ScanCache = scan_cache;
+
+        ScanOutput scan_output;
+
+        if (ScanImplicitDeps(stat_cache, &scan_input, &scan_output))
+        {
+          for (int i = 0, count = scan_output.m_IncludedFileCount; i < count; ++i)
+          {
+            const FileAndHash& path = scan_output.m_IncludedFiles[i];
+            if (HashTableLookup(&implicitDependencies, path.m_FilenameHash, path.m_Filename) == nullptr)
+              HashTableInsert(&implicitDependencies, path.m_FilenameHash, path.m_Filename, false);
+          }
+        }
+      }
+
+      bool implicitFilesListChanged = implicitDependencies.m_RecordCount != prev_state->m_ImplicitInputFiles.GetCount();
+      if (!implicitFilesListChanged)
+      {
+        for (const NodeInputFileData& implicitInput : prev_state->m_ImplicitInputFiles)
+        {
+          bool* visited = HashTableLookup(&implicitDependencies, Djb2HashPath(implicitInput.m_Filename), implicitInput.m_Filename);
+          if (!visited)
+          {
+            implicitFilesListChanged = true;
+            break;
+          }
+
+          *visited = true;
+        }
+
+        HashTableWalk(&implicitDependencies, [&](int32_t index, uint32_t hash, const char* filename, bool visited)
+        {
+          if (!visited)
+            implicitFilesListChanged = true;
+        });
+      }
+
+      if (implicitFilesListChanged)
+      {
+        JsonWriteStartObject(msg);
+
+        JsonWriteKeyName(msg, "key");
+        JsonWriteValueString(msg, "InputFileList");
+
+        JsonWriteKeyName(msg, "value");
+        JsonWriteStartArray(msg);
+        for (const FrozenFileAndHash& input : node_data->m_InputFiles)
```

Comment:
This is the wrong file list to be reporting... we need to collect the scanner results, and use prev_state->m_ImplicitInputFiles, below.

Reply:
Fixed in c82ed7dccefd67f5d2b07694cefab09fdc8285f0